### ISSUE TITLE
BeginTx missing constraints

### ISSUE THREAD

**Issue by ed255** [privacy-scaling-explorations/zkevm-circuits#1475]
Previously we identified that the BeginTx circuit implementation had constraints for the `is_create` case that were missing in the specs https://github.com/privacy-scaling-explorations/zkevm-specs/issues/420

Adria has been working on syncing the specs with the circuit and noticed that there are missing constraints in the circuits!

I think the best way to approach this is
- first: work on https://github.com/privacy-scaling-explorations/zkevm-specs/issues/420 and complete the spec of BeginTx (and consider all missing stuff)
- second: update the circuit to match the spec

Example of missing constraints:
- The code_hash set in the step state transition should be constrained to be the hash of the tx.calldata:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b06f44f4609487bc5bc78179c42fe4093e2e1307/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L285
- The callee address of the new context should be the `keccak(RLP([sender_address, sender_nonce]))`:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b06f44f4609487bc5bc78179c42fe4093e2e1307/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L269

**Comment by ChihChengLiang** [comment]
This is unblocked.

**Comment by davidnevadoc** [comment]
As @KimiWu123 suggested in Discord, it is a good idea to postpone this task until spec and circuit are in sync. 
The tasks/ issues that should be resolved first are: 
 - https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1526
 - https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1611

**Comment by curryrasul** [comment]
Hey @ed255 , may I ask please, you said:
> The code_hash set in the step state transition should be constrained to be the hash of the tx.calldata

Can it only be equal to hash of tx.calldata? Cause in case of internal call contract creation it'll be equal to hash of the bytecode that's stored in memory. Here's the updated link to code:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8d230b348e8eca19f0a80d617920efb714b7be34/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L157

**Comment by ed255** [comment]
> Hey @ed255 , may I ask please, you said:
> 
> > The code_hash set in the step state transition should be constrained to be the hash of the tx.calldata
> 
> Can it only be equal to hash of tx.calldata? 

The check must be conditional to the contract creation transaction case.  In such a case, the executed code is the one found in `tx.calldata`, so we should set `context.code_hash = hash(tx.calldata)`, so that every time we look up a byte of the code (which is indexed by `code_hash`) it is bound to `tx.calldata`.

> Cause in case of internal call contract creation it'll be equal to hash of the bytecode that's stored in memory. Here's the updated link to code:
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8d230b348e8eca19f0a80d617920efb714b7be34/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L157

Hm, but "internal call contract creation" is not handled via `begin_tx` right?  That would be in the `CallOpGadget` I think.

From the updated code you shared, in the contract creation case:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8d230b348e8eca19f0a80d617920efb714b7be34/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L295
But I think `cb.curr.state.code_hash` is not constrained at, it should be constrainted to be `= hash(tx.calldata)`

**Comment by ChihChengLiang** [comment]
Hey @ed255,
The constraint `context.code_hash = hash(tx.calldata)` seems to be incorrect.
The code corresponding to the hash should be the output of the constructor.

Please see [this example](https://www.evm.codes/playground?callValue=9&unit=Wei&codeType=Mnemonic&code=%27z0q0f9q9f0y4+FFmslk3+0x63FFFFFFFF6000526004601CF3gvMSTORE%7E13%7E19gp+%27%7Ek+z%2F%2F+Createmnmccountjith+yjeimnd+v%5Cnqynolgg%7EpvCREATEm+al+codekvPUSH1j+wg%7E0fpvvz%01fgjklmpqvyz%7E_&fork=shanghai
), we run `CREATE` opcode to deploy `FFFFFFFF` as the byte code. The `CREATE` yields the deployed address as output, we then call `EXTCODEHASH` to extract the code hash from it. The resulting code hash is `2904...c238`, which is exactly [`hash("FFFFFFFF")`](https://emn178.github.io/online-tools/keccak_256.html?input_type=hex&input=FFFFFFFF)

Thank @zemse for pointing out this in a call.

**Comment by ChihChengLiang** [comment]
Edu clarified in a call.
So begin tx has a implicit call context and a call context must have a code hash. The reason for adding the code hash is for the circuit but not for the EVM. 

The code hash is for the circuit to constraint the content of init_code. The EVM doesn't require this step.
